File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap14/ThreadInterference.py
#!/usr/bin/env python

###  ThreadInterference.py

import threading       
import time

import random                  
ran = random.Random()          

#-------------------------   class DataObject  ------------------------- 
class DataObject( object ):                                          #(A)
    def __init__( self ):                                            #(B)
        self.dataItem1 = 50                                          #(C)
        self.dataItem2 = 50                                          #(D)

    def itemSwap(self):                                              #(E)
        x = -5 + ran.randint(0,10)                                   #(F)
        self.dataItem1 -= x                                          #(G)
        self.keepBusy()                                              #(H)
        self.dataItem2 += x                                          #(I)

    def test(self):                                                  #(J)
        sum = self.dataItem1 + self.dataItem2                        #(K)
        print "Sum: %d" % sum                                        #(L)

    def keepBusy( self ):                                            #(M)
        for i in range(100000): pass                                 #(N)

#---------------------   class RepeatedSwaps   -------------------------
class RepeatedSwaps( threading.Thread ):                             #(O)
    def __init__( self, dataObject ):                                #(P)
        self.dataObject = dataObject                                 #(Q)
        threading.Thread.__init__(self)                              #(R)

    def run( self ):                                                 #(S)
        for i in range (500):                                        #(T)
            self.dataObject.itemSwap()                               #(U)
            if i % 100 == 0:                                         #(V)
                self.dataObject.test()                               #(W)

#---------------------------   launch threads -------------------------

d = DataObject()                                                     #(X)

t1 = RepeatedSwaps(d)                                                #(Y)
t2 = RepeatedSwaps(d)                                                #(Z)
t3 = RepeatedSwaps(d)                                                #(a)

map( lambda x: x.start(), (t1, t2, t3) )                             #(b)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap14/ThreadInterference.pl
#!/usr/bin/perl -w

###  ThreadInterference.pl

use strict;
use threads;                                                         #(A)
use threads::shared;                                                 #(B)

my @dataObj : shared = (50, 50);                                     #(C)

my $t1 = threads->create( \&repeatedSwaps );                         #(D)
my $t2 = threads->create( \&repeatedSwaps );                         #(E)
my $t3 = threads->create( \&repeatedSwaps );                         #(F)

$t1->join;                                                           #(G)
$t2->join;                                                           #(H)
$t3->join;                                                           #(I)

sub itemSwap {                                                       #(J)
    my $x = int ( -4.999999 + rand(10 ) );                           #(K)
    $dataObj[0] += $x;                                               #(L)
    keepBusy();                                                      #(M)
    $dataObj[1] -= $x;                                               #(N)
}

sub keepBusy {                                                       #(O)
    my $i = 0;                                                       #(P)
    while ( $i++ < 10000 ) {}                                        #(Q)
}

sub test {                                                           #(R)
    my $sum = $dataObj[0] + $dataObj[1];                             #(S)
    print "Sum: $sum\n";                                             #(T)
}

sub repeatedSwaps {                                                  #(U)
    my $j = 0;                                                       #(V)
    while ( $j++ < 500 ) {                                           #(W)
        itemSwap();                                                  #(X)
        test() unless $j % 100;                                      #(Y)
    }
}
